File: Medium/547. Friend Circles/solution.java
/**
 * DFS in graph
 */

class Solution {
    public int findCircleNum(int[][] M) {
        int len = M.length;
        boolean[] visited = new boolean[len];
        int count = 0;
        for (int i = 0; i < len; i++) {
            if (!visited[i]) {
                dfs(M, visited, i);
                count++;
            }
        }
        return count;
    }

    private static void dfs(int[][] M, boolean[] visited, int i) {

        visited[i] = true;
        int len = M[i].length;
        for (int j = 0; j < len; j++) {
            if (M[i][j] == 1 && !visited[j]) {
                dfs(M, visited, j);
            }
        }
    }
}

File: Medium/547. Friend Circles/solution.py
class Solution:
    def findCircleNum(self, M: List[List[int]]) -> int:
        
        def dfs(person):
            M[person][person] = 0
            for i in range(rows):
                if M[i][i] != 0 and M[person][i] == 1:
                    dfs(i)
        rows = len(M)
        count = 0
        for i in range(rows):
            if M[i][i] != 0:
                dfs(i)
                count += 1
        
        return count